src/main/java/com/example/SS2_Backend/model/stableMatching/StableMatchingOTOProblem.java

```diff
@@ -0,0 +1,208 @@
+                preferences[i] = individualPref.getPositions();
+            } else {
+                preferences[i] = addPadding(individualPref.getPositions(), this.padding);
+            }
+        }
+    }
+    static int[] addPadding(int[] array, int padding) {
+        for (int i = 0; i < array.length; i++) {
+            array[i] += padding;
+        }
+        return array;
+    }
+    public PreferenceList getPreferenceOfIndividual(int index) {
+        PreferenceList a;
+        if (!func1 && !func2) {
+            a = preferencesProvider.getPreferenceListByDefault(index);
+        } else {
+            a = preferencesProvider.getPreferenceListByFunction(index);
+        }
+        return a;
+    }
+
+    private void initializePrefProvider() {
+        if (this.evaluateFunctionForSet1 != null) {
+            this.preferencesProvider.setEvaluateFunctionForSet1(evaluateFunctionForSet1);
+        }
+        if (this.evaluateFunctionForSet2 != null) {
+            this.preferencesProvider.setEvaluateFunctionForSet2(evaluateFunctionForSet2);
+        }
+    }
+
+    public void setEvaluateFunctionForSet1(String func) {
+        if(VALID_EVALUATE_FUNCTION_KEYWORDS.stream().anyMatch(func::contains)) {
+            this.func1 = true;
+            this.evaluateFunctionForSet1 = func;
+        }
+    }
+    public void setEvaluateFunctionForSet2(String func) {
+        if(VALID_EVALUATE_FUNCTION_KEYWORDS.stream().anyMatch(func::contains)) {
+            this.func2 = true;
+            this.evaluateFunctionForSet2 = func;
+        }
+    }
+    public void setFitnessFunction(String func) {
+        if (func.contains("S") || func.contains("SIGMA{") ||
+                func.contains("M")) {
+            this.fitfunc = true;
+            this.fitnessFunction = func;
+        }
+    }
+
+    @Override
+    public String getName() {
+        return this.getClass().getSimpleName();
+    }
+    @Override
+    public int getNumberOfVariables() {
+        return 1;
+    }
+    @Override
+    public int getNumberOfObjectives() {
+        return 1;
+    }
+    @Override
+    public int getNumberOfConstraints() {
+        return 0;
+    }
+    @Override
+    public void evaluate(Solution solution) {
+        log.info("Evaluating ... ");
+        int[] order = ((Permutation) solution.getVariable(0)).toArray();
+        MatchesOTO matches = StableMatchingAlgorithm(order);
+        double[] satisfaction = getAllSatisfactions(matches);
+        solution.setAttribute("matches", matches);
+        if (satisfaction.length == 0) {
+            solution.setObjective(0, 1);
+            log.info("Skip infinite loop");
+        } else {
+            double sum = Arrays.stream(satisfaction).sum();
+            solution.setObjective(0, -sum);
+            log.info("Score: {}", convertToStringWithoutScientificNotation(sum));
+        }
+    }
+    public MatchesOTO StableMatchingAlgorithm(int[] order) {
+        Queue<Integer> singleQueue = new LinkedList<>();
```

Comment:
xem xét cách Queue<Integer> queue = new LinkedList<>(Arrays.asList(array)); xem có được không, hạn chế for loop